File: app/lib/syllabify.ts
// eslint-disable-next-line @typescript-eslint/no-require-imports
const hyphenModules: Record<string, { hyphenateSync: (w: string) => string }> = {
  it: require("hyphen/it"),
  en: require("hyphen/en"),
  "en-us": require("hyphen/en-us"),
  "en-gb": require("hyphen/en-gb"),
  es: require("hyphen/es"),
  fr: require("hyphen/fr"),
  de: require("hyphen/de"),
  pt: require("hyphen/pt"),
  ca: require("hyphen/ca"),
  nl: require("hyphen/nl"),
  pl: require("hyphen/pl"),
  ru: require("hyphen/ru"),
  sv: require("hyphen/sv"),
  nb: require("hyphen/nb"),
  da: require("hyphen/da"),
  fi: require("hyphen/fi"),
  ro: require("hyphen/ro"),
  cs: require("hyphen/cs"),
  sk: require("hyphen/sk"),
  hr: require("hyphen/hr"),
  tr: require("hyphen/tr"),
  eu: require("hyphen/eu"),
  gl: require("hyphen/gl"),
  af: require("hyphen/af"),
};

// Whisper outputs ISO 639-1 codes; map variants and aliases
const LANG_ALIASES: Record<string, string> = {
  "en-us": "en-us",
  "en-gb": "en-gb",
  "zh-tw": "en", // fallback
  "pt-br": "pt",
  "pt-pt": "pt",
  "de-at": "de",
  "de-ch": "de",
  "fr-ca": "fr",
  "fr-be": "fr",
  "nl-be": "nl",
  "sr-latn": "en",  // fallback
  no: "nb",
};

const SOFT_HYPHEN = "­";

function resolveHyphenator(lang: string): { hyphenateSync: (w: string) => string } | null {
  const key = (LANG_ALIASES[lang] ?? lang).toLowerCase();
  return hyphenModules[key] ?? hyphenModules[key.split("-")[0]] ?? null;
}

/**
 * Splits a single word into syllables using TeX hyphenation patterns.
 * Falls back to [word] for unsupported languages or words ≤ 3 chars.
 */
export function splitWord(word: string, lang: string): string[] {
  const clean = word.trim();
  if (!clean || clean.length <= 2) return [clean].filter(Boolean);

  const h = resolveHyphenator(lang);
  if (!h) return [clean];

  try {
    const hyphenated = h.hyphenateSync(clean);
    const parts = hyphenated.split(SOFT_HYPHEN).filter(Boolean);
    // hyphenateSync may return the original word unchanged if no break points found
    return parts.length > 0 ? parts : [clean];
  } catch {
    return [clean];
  }
}

/** Splits all words in a line into syllable arrays. */
export function syllabifyLine(
  line: string,
  lang: string
): Array<{ word: string; syllables: string[] }> {
  return line
    .trim()
    .split(/\s+/)
    .filter(Boolean)
    .map((word) => ({ word, syllables: splitWord(word, lang) }));
}

File: cli/syllabify.py
"""Syllabification using pyphen (replaces hyphen npm package)."""

import pyphen

from cli.logging_setup import get_logger

logger = get_logger("cli.syllabify")

# Whisper ISO 639-1 codes -> pyphen locale codes
_LANG_ALIASES: dict[str, str] = {
    "en": "en_US",
    "en-us": "en_US",
    "en-gb": "en_GB",
    "zh": "en_US",    # fallback — pyphen has no Chinese
    "zh-tw": "en_US",
    "pt": "pt_BR",
    "pt-br": "pt_BR",
    "pt-pt": "pt_BR",
    "de": "de_DE",
    "de-at": "de_DE",
    "de-ch": "de_DE",
    "fr": "fr_FR",
    "fr-ca": "fr_FR",
    "fr-be": "fr_FR",
    "es": "es_ES",
    "it": "it_IT",
    "nl": "nl_NL",
    "nl-be": "nl_NL",
    "pl": "pl_PL",
    "ru": "ru_RU",
    "sv": "sv_SE",
    "nb": "nb_NO",
    "no": "nb_NO",
    "da": "da_DK",
    "fi": "fi_FI",
    "ro": "ro_RO",
    "cs": "cs_CZ",
    "sk": "sk_SK",
    "hr": "hr_HR",
    "tr": "tr_TR",
    "ca": "ca_ES",
    "gl": "gl_ES",
    "sr-latn": "en_US",  # fallback
}

# Cached hyphenator instances
_hyphenators: dict[str, pyphen.Pyphen] = {}


def _get_hyphenator(lang: str) -> pyphen.Pyphen | None:
    """Get or create a pyphen hyphenator for the given language."""
    locale = _LANG_ALIASES.get(lang, lang.split("-")[0])
    if locale not in _hyphenators:
        try:
            _hyphenators[locale] = pyphen.Pyphen(lang=locale)
        except (ValueError, KeyError):
            # Language not available in pyphen
            _hyphenators[locale] = None  # type: ignore[assignment]
    return _hyphenators[locale]


def split_word(word: str, lang: str) -> list[str]:
    """Split a single word into syllables using pyphen.

    Args:
        word: The word to split.
        lang: ISO 639-1 language code.

    Returns:
        List of syllable strings. For words <= 2 chars or unsupported
        languages, returns the whole word as a single syllable.
    """
    clean = word.strip()
    if not clean or len(clean) <= 2:
        return [clean] if clean else []

    h = _get_hyphenator(lang)
    if h is None:
        return [clean]

    try:
        parts = h.divided(clean)
        parts = [p for p in parts if p]
        return parts if parts else [clean]
    except Exception:
        return [clean]


def syllabify_line(line: str, lang: str) -> list[list[str]]:
    """Split all words in a line into syllable arrays.

    Args:
        line: The text line to syllabify.
        lang: ISO 639-1 language code.

    Returns:
        List of syllable lists, one per word.
    """
    return [
        split_word(word, lang)
        for word in line.strip().split()
        if word
    ]
